Add "What's new" for C# 8 preview 2 (#10628)
dotnet/docs: docs/csharp/whats-new/csharp-8.md

title: What's New in C# 8.0 - C# Guide
description: Get an overview of the new features available in C# 8.0. This article is up-to-date with preview 2.
ms.date: 02/12/2019
---
# What's new in C# 8.0

> [!NOTE]
> This article was last updated for C# 8.0 preview 2.

There are many enhancements to the C# language that you can try out already with preview 2. The new features added in preview 2 are:

- [Pattern matching enhancements](#more-patterns-in-more-places):
  * [Switch expressions](#switch-expressions)
  * [Property patterns](#property-patterns)
  * [Tuple patterns](#tuple-patterns)
  * [Positional patterns](#positional-patterns)
- [Using declarations](#using-declarations)
- [Static local functions](#static-local-functions)
- [Disposable ref structs](#disposable-ref-structs)

The following language features first appeared in C# 8.0 preview 1:

- [Nullable reference types](#nullable-reference-types)
- [Asynchronous streams](#asynchronous-streams)
- [Ranges and indices](#ranges-and-indices)

The remainder of this article briefly describes these features. Where in-depth articles are available, links to those tutorials and overviews are provided.

## More patterns in more places

> **Pattern matching** gives tools to provide shape-dependent functionality across related but different kinds of data. C# 7.0 introduced syntax for type patterns and constant patterns by using the `is` expression and the `switch` statement. These features represented the first tentative steps toward supporting programming paradigms where data and functionality live apart. As the industry moves toward more microservices and other cloud-based architectures, other language tools are needed.

Author: I don't think perf work has been completed. So I won't address this yet. I think it will be similar perf, but I'll verify as we get closer to release.

### switch expressions

Often, a switch statement produces a value in each of its `case` blocks. **Switch expressions** enable you to use more concise expression syntax. There are fewer repetitive `case`, and `break` keywords, and fewer curly braces.  As an example, consider the following enum that lists the colors of the rainbow:

```csharp
public enum Rainbow
{
    Red,
    Orange,
    Yellow,
    Blue,
    Indigo,
    Violet
}
```

You could convert a `Rainbow` value to its RGB values using the following method containing a switch expression:

```csharp
public static RGBColor fromRainbow(Rainbow colorBand) =>
    colorBand switch
    {
        Rainbow.Red    => new RGBColor(0xFF, 0x00, 0x00),
        Rainbow.Orange => new RGBColor(0xFF, 0x7F, 0x00),
        Rainbow.Yellow => new RGBColor(0xFF, 0xFF, 0x00),
        Rainbow.Blue   => new RGBColor(0x00, 0x00, 0xFF),
        Rainbow.Indigo => new RGBColor(0x4B, 0x00, 0x82),
        Rainbow.Violet => new RGBColor(0x94, 0x00, 0xD3),
        _              => throw new ArgumentException(message: "invalid enum value", paramName: nameof(colorBand)),
    };
```

There are several syntax improvements here.

> - The variable comes before the `switch` keyword. That enables switch expression to compose using recursive patterns. The different order makes it visually easy to distinguish from the switch statement.

Author: I removed that sentence here. I added a recursive example later on.

### Positional patterns

Some types include a `Deconstruct` method that deconstructs its properties into discrete variables. When a `Deconstruct` method is accessible, you can use **positional patterns** to inspect properties of the object and use those properties for a pattern.  Consider the following `Point` class that includes a `Deconstruct` method to create discrete variables for `X` and `Y`:

```csharp
public class Point
{
    public int X { get; }
    public int Y { get; }

    public Point(int x, int y) => (X, Y) = (x, y);

    public void Deconstruct(out int x, out int y) =>
        (x, y) = (X, Y);
}
```

The following method uses the **positional pattern** to extract the values of `x` and `y`. Then, it uses a `when` clause to determine the quadrant of the point:

```csharp
static string Quadrant(Point p) => p switch
{
    (0, 0) => "origin",
    (var x, var y) when x > 0 && y > 0 => "Quadrant 1",
    (var x, var y) when x < 0 && y > 0 => "Quadrant 2",
    (var x, var y) when x < 0 && y < 0 => "Quadrant 3",
    (var x, var y) when x > 0 && y < 0 => "Quadrant 4",
    (var x, var y) => "on a border",

>     _ => "unknown"

Author: It's triggered when either X or Y (but not both) is 0.

};
```

## using declarations

A **using declaration** is a variable declaration preceded by the `using` keyword. It tells the compiler that the variable being declared should be disposed at the end of the enclosing scope. For example, consider the following code that writes a text file:

> ```csharp

Reviewer: In the previous pattern matching section you used the format:
```
NEW

compare to

OLD
```

but in this one you are starting with the old way of doing it and showing the new way. I would rather see the new way upfront because most likely everyone is familiar with how to write a `using` statement and would rather just see the change first.

    {
        foreach (string line in lines)
        {
            // If the line doesn't contain the word 'Second', write the line to the file.
            if (!line.Contains("Second"))
            {
                file.WriteLine(line);
            }
        }
    } // file is disposed here
}
```

A few braces are removed by replacing the `using` statement with a `using` declaration:

```csharp
static void WriteLinesToFile(IEnumerable<string> lines)
{
    using var file = new System.IO.StreamWriter("WriteLines2.txt");
    foreach (string line in lines)
    {
        // If the line doesn't contain the word 'Second', write the line to the file.
        if (!line.Contains("Second"))
        {
            file.WriteLine(line);
        }
    }
// file is disposed here
}
```

Using declarations provide the same behavior when exceptions are thrown: the `Dispose()` call is in a `finally` block.s

## Static local functions

You can now add the `static` modifier to local functions to ensure that local function doesn't capture (reference) any variables from the enclosing scope. Doing so generates `CS8421`, "A static local function can't contain a reference to <variable>." 

> Consider the following code. The local function `LocalFunction` accesses the variable `y`, declared in the enclosing scope (the method `M`). Therefore, `LocalFunction` can't be declared with the `static` modifier:

Reviewer: I agree with this comment. One place I used a local function was to do a comparison on a backing `List`. I had a `AddObject` type of method that looked at the object and added it to one of four lists and then called `List.Sort` passing in the local function to do the compare. In C#8 that will now be a static function since it was a compare method

```csharp
int CompareComponent(IComponent left, IComponent right)
{
    if (left.SortOrder > right.SortOrder)
        return 1;

    if (left.SortOrder < right.SortOrder)
        return -1;

    return 0;
}
```

> Consider the following code. The local function `LocalFunction` accesses the variable `y`, declared in the enclosing scope (the method `M`). Therefore, `LocalFunction` can't be declared with the `static` modifier:

Author: I went with a simpler example for this topic. When I write the "explore" topic, with interactive samples, this is a good one.

```csharp
int M()
{
    int y;
    LocalFunction();
    return y;

    void LocalFunction() => y = 0;
}
```

## Disposable ref structs

> A `struct` declared with the `ref` modifier may not implement any interfaces and so cannot implement <xref:System.IDisposable>. Therefore, to enable a `ref struct` to be disposed, it must have an accessible `void Dispose()` method.

Reviewer: I would clarify how `void Dispose()` gets called. Is this something that has to be done by hand? Will the runtime do it for you? How does it interact with using statements? In C# 7 you get an error if you try to use an object that doesn't implement IDisposable: `using (var a = new ObjNotImplementIDisposable)`: did that change with C# 8?

## Nullable reference types

Inside a nullable annotation context, any variable of a reference type is considered to be a **nonnullable reference type**. If you want to indicate that a variable may be null, you must append the type name with the `?` to declare the variable as a **nullable reference type**.

> For nonnullable reference types, the compiler uses flow analysis to ensure that local variables are initialized to a non-null value when declared. Fields must be initialized to a non-null value using either an initializer or all constructors. Furthermore, nonnullable reference types can't be assigned a value that could be null.

Author: I reworked the sentence.

> For nonnullable reference types, the compiler uses flow analysis to ensure that local variables are initialized to a non-null value when declared. Fields must be initialized to a non-null value using either an initializer or all constructors. Furthermore, nonnullable reference types can't be assigned a value that could be null.

Reviewer: I think it should be `in all constructors`

## Asynchronous streams

Starting with C# 8.0, you can create and consume streams asynchronously. A method that returns an asynchronous stream has three properties:

1. It was declared with the `async` modifier.
1. It returns an `IAsyncEnumerable<T>`.
1. The method contains `yield return` statements to return successive elements in the asynchronous stream.

Consuming an asynchronous stream requires you to add the `await` keyword before the `foreach` keyword when you enumerate the elements of the stream. Adding the `await` keyword requires the method that enumerates the asynchronous stream was declared with the `async` modifier and returns a type allowed for an `async` method. Typically that means returning a <xref:System.Threading.Tasks.Task> or <xref:System.Threading.Tasks.Task%601>. It can also be a <xref:System.Threading.Tasks.ValueTask> or <xref:System.Threading.Tasks.ValueTask%601>. A method can both consume and produce an asynchronous stream, which means it would return an `IAsyncEnumerable<T>`.

You can try asynchronous streams yourself in our tutorial on [creating and consuming async streams](../tutorials/generate-consume-asynchronous-stream.md).

> ## Ranges and indices

Reviewer 2: This is a really useful feature, but I just don't get the syntax. It would be helpful to briefly describe the syntax of ranges and indices rather than just provide a set of examples. (Unless I'm just being obtuse...)

Author: I rewrote this entire section.  I do think I made it much harder to understand than it is.

> Ranges and indices provide a succinct syntax for specifying subranges in an array, <xref:System.Span%601>, or <xref:System.ReadonlySpan%601>. Consider the following array:

Reviewer 2: ```suggestion
Ranges and indices provide a succinct syntax for specifying subranges in an array, <xref:System.Span%601>, or <xref:System.ReadOnlySpan%601>. Consider the following array:
```

```csharp
var words = new string[]
{
    "The",
    "quick",
    "brown",
    "fox",
    "jumped",
    "over",
    "the",
    "lazy",
    "dog"
};
```

You can retrieve the last word by prefixing the index `1` with `^`:

```csharp
Console.WriteLine($"The last word is {words[^1]}");
```

When you take only a single element and specify from the end, indexing starts at 1.

You can also specify ranges of an array. The following code creates a subrange with the words "quick", "brown", and "fox":

```csharp

> var brownFox = words[1..4];

Reviewer 2: It seems confusing that this returns only 3 words, when you'd expect it to return the elements at indices 1 through 4, or 4 words An explanation of the syntax would help.

> When you take only a single element and specify from the end, indexing starts at 1.

Reviewer: I would state this above the code with the instructions on retrieving the last word.

```

Ranges can be specified from the end as well as the beginning. The following code creates a subrange with "lazy" and "dog":

```csharp

> var lazyDog = words[^2..^0];

Reviewer 2: Also confusing. If indexing from the last element is one-based, why isn't this `^2..^1`?

```

> You can also create ranges that are open ended for the start, end, or both, as in the following code:

Reviewer: It may be easier to start with this stuff because this is a little more familiar to those used to 0-based indexing. You may want to start with using the `Range` type directly and then say that there is some sort of implicit conversion going on with the range syntax? (if that is true, I just assume it is)

**Pattern matching** gives tools to provide shape-dependent functionality across related but different kinds of data. C# 7.0 introduced syntax for type patterns and constant patterns by using the [`is`](../language-reference/keywords/is.md) expression and the [`switch`](../language-reference/keywords/switch.md) statement. These features represented the first tentative steps toward supporting programming paradigms where data and functionality live apart. As the industry moves toward more microservices and other cloud-based architectures, other language tools are needed.

C# 8.0 expands this vocabulary so you can use more pattern expressions in more places in your code. Consider these features when your data and functionality are separate. Consider pattern matching when your algorithms depend on a fact other than the runtime type of an object. These techniques provide another way to express designs.

In addition to new patterns in new places, C# 8.0 adds **recursive patterns**. The result of any pattern expression is an expression. A recursive pattern is simply a pattern expression applied to the output of another pattern expression.

### switch expressions

> Often, a switch statement produces a value in each of its `case` blocks. **Switch expressions** enable you to use more concise expression syntax. There are fewer repetitive `case`, and `break` keywords, and fewer curly braces.  As an example, consider the following enum that lists the colors of the rainbow:

Reviewer 2: ```suggestion
Often, a switch statement produces a value in each of its `case` blocks. **Switch expressions** enable you to use more concise expression syntax. There are fewer repetitive `case` and `break` keywords, and fewer curly braces.  As an example, consider the following enum that lists the colors of the rainbow:
```

```csharp
public enum Rainbow
{
    Red,
    Orange,
    Yellow,
    Blue,
    Indigo,
    Violet
}
```

You could convert a `Rainbow` value to its RGB values using the following method containing a switch expression:

```csharp
public static RGBColor FromRainbow(Rainbow colorBand) =>
    colorBand switch
    {
        Rainbow.Red    => new RGBColor(0xFF, 0x00, 0x00),
        Rainbow.Orange => new RGBColor(0xFF, 0x7F, 0x00),
        Rainbow.Yellow => new RGBColor(0xFF, 0xFF, 0x00),
        Rainbow.Blue   => new RGBColor(0x00, 0x00, 0xFF),
        Rainbow.Indigo => new RGBColor(0x4B, 0x00, 0x82),
        Rainbow.Violet => new RGBColor(0x94, 0x00, 0xD3),
        _              => throw new ArgumentException(message: "invalid enum value", paramName: nameof(colorBand)),
    };
```

There are several syntax improvements here:

- The variable comes before the `switch` keyword. The different order makes it visually easy to distinguish the switch expression from the switch statement.

> - The `case` and `:` elements are replaced with `=>`. It's more concise.

Reviewer 2: ```suggestion
- The `case` and `:` elements are replaced with `=>`. It's more concise and intuitive.
```

        if (!line.Contains("Second"))
        {
            file.WriteLine(line);
        }
    }
// file is disposed here
}
```

In the preceding example, the file is disposed when the closing brace for the method is reached. That's the end of the scope in which `file` is declared. The preceding code is equivalent to the following code using the classic [using statements](../language-reference/keywords/using-statement.md) statement:

```csharp
static void WriteLinesToFile(IEnumerable<string> lines)
{
    using (var file = new System.IO.StreamWriter("WriteLines2.txt"))
    {
        foreach (string line in lines)
        {
            // If the line doesn't contain the word 'Second', write the line to the file.
            if (!line.Contains("Second"))
            {
                file.WriteLine(line);
            }
        }
    } // file is disposed here
}
```

In the preceding example, the file is disposed when the closing brace associated with the `using` statement is reached.

In both cases, the compiler generates the call to `Dispose()`. The compiler generates an error if the expression in the using statement is not disposable.

## Static local functions

> You can now add the `static` modifier to local functions to ensure that local function doesn't capture (reference) any variables from the enclosing scope. Doing so generates `CS8421`, "A static local function can't contain a reference to <variable>." 

Reviewer 2: ```suggestion
You can now add the `static` modifier to local functions to ensure that a local function doesn't capture (reference) any variables from the enclosing scope. Doing so generates `CS8421`, "A static local function can't contain a reference to <variable>." 
```

```csharp
public static async System.Collections.Generic.IAsyncEnumerable<int> GenerateSequence()
{
    for (int i = 0; i < 20; i++)
    {
        await Task.Delay(100);
        yield return i;
    }
}
```

You would enumerate the sequence using the `await foreach` statement:

```csharp
await foreach (var number in GenerateSequence())
{
    Console.WriteLine(number);
}
```

You can try asynchronous streams yourself in our tutorial on [creating and consuming async streams](../tutorials/generate-consume-asynchronous-stream.md).

## Indices and ranges

Ranges and indices provide a succinct syntax for specifying subranges in an array, <xref:System.Span%601>, or <xref:System.ReadOnlySpan%601>.

You can specify an index **from the end**. You specify **from the end** using the `^` operator. You are familiar with `array[2]` meaning the element "2 from the start". Now, `array[^2]` means the element "2 from the end". The index `^0` means "the end", or one past that last element.

> You can specify a **range** with the **range operator**: `..`. For example, `0..^0` specifies the entire range of the array: 0 from the start up to, but not including 0 from the end. Either operand may use "from the start" or "from the end". Furthermore, either operand may be omitted. The defaults are `0` for the start index, and `^0` for the end index.

Reviewer: It might be worth pointing out that the end of a range is exclusive of the index. Much like how random works. The start is inclusive lower bound of the value, but the end is exclusive upper bound of the value. (I'm assuming this is how range works?)

And additionally, it seem the `^` (from end) indicator is also exclusive upper bound-based.

The "2 from start" doesn't make sense to me either. I would make sure to use **element-index** or something to clarify that. When you say `element "2 from the start"` it sounds like you're pulling the element value of 2 out, versus `element index 2 from the start` or `3rd element from the start` would be clearer. Or compare it more directly and indicate the difference with something like:

> `array[2]` means the 3rd element from the start of the array and `array[^2]` means the 2nd element from the end of the array.

And how can you go passed the end of an array with `^0` it seems like that should be invalid? I've not really investigated this stuff so these are my "first look" thoughts :)

Let's look at a few examples. Consider the following array:

```csharp
var words = new string[]
{

>     "The",

Reviewer: would it be useful to put in the element numbers here for index vs ending index?

```csharp
var words = new string[]
{
                // index    ending index
    "The",      // 0        ^9
    "quick",    // 1        ^8
    "brown",    // 2        ^7
    "fox",      // 3        ^6
    "jumped",   // 4        ^5
    "over",     // 5        ^4
    "the",      // 6        ^3
    "lazy",     // 7        ^2
    "dog"       // 8        ^1
};
```

                file.WriteLine(line);
            }
        }
    } // file is disposed here
}
```

In the preceding example, the file is disposed when the closing brace associated with the `using` statement is reached.

In both cases, the compiler generates the call to `Dispose()`. The compiler generates an error if the expression in the using statement is not disposable.

## Static local functions

You can now add the `static` modifier to local functions to ensure that local function doesn't capture (reference) any variables from the enclosing scope. Doing so generates `CS8421`, "A static local function can't contain a reference to <variable>." 

Consider the following code. The local function `LocalFunction` accesses the variable `y`, declared in the enclosing scope (the method `M`). Therefore, `LocalFunction` can't be declared with the `static` modifier:

```csharp
int M()
{
    int y;
    LocalFunction();
    return y;

    void LocalFunction() => y = 0;
}
```

The following code contains a static local function. It can be static because it doesn't access any variables in the enclosing scope:

```csharp
int M()
{
    int y = 5;
    int x = 7;
    return Add(x, y);

>     int Add(int left, int right) => left + right;

Reviewer 3: @BillWagner is the local function intentionally not `static` in this example, though it can be such? For the sake of example, adding `static` modifier is reasonable.

}
```

## Disposable ref structs

> A `struct` declared with the `ref` modifier may not implement any interfaces and so cannot implement <xref:System.IDisposable>. Therefore, to enable a `ref struct` to be disposed, it must have an accessible `void Dispose()` method.

Reviewer 3: I guess `readonly ref struct` can also be disposable as a particular case of `ref struct`?

## Nullable reference types

Inside a nullable annotation context, any variable of a reference type is considered to be a **nonnullable reference type**. If you want to indicate that a variable may be null, you must append the type name with the `?` to declare the variable as a **nullable reference type**.

> For nonnullable reference types, the compiler uses flow analysis to ensure that local variables are initialized to a non-null value when declared. Fields must be initialized during construction. The compiler generates a warning if the variable is not set in all constructors, or an initializer. Furthermore, nonnullable reference types can't be assigned a value that could be null.

Reviewer 3: ```suggestion
For nonnullable reference types, the compiler uses flow analysis to ensure that local variables are initialized to a non-null value when declared. Fields must be initialized during construction. The compiler generates a warning if the variable is not set in any of available constructors, or an initializer. Furthermore, nonnullable reference types can't be assigned a value that could be null.
```

Nullable reference types aren't checked to ensure they aren't assigned or initialized to null. However, the compiler uses flow analysis to ensure that any variable of a nullable reference type is checked against null before it's accessed or assigned to a nonnullable reference type.

> You can learn more about the feature in the overview of [nullable reference types](../nullable-references.md). Try it yourself in a new application in this [nullable reference types tutorial](../tutorials/nullable-reference-types.md). Learn about the steps to migrate an existing codebase to make use of nullable reference types in the [migrating an application to use nullable reference types tutorial]../tutorials/upgrade-to-nullable-references.md).

Reviewer 3: Broken link:
```suggestion
You can learn more about the feature in the overview of [nullable reference types](../nullable-references.md). Try it yourself in a new application in this [nullable reference types tutorial](../tutorials/nullable-reference-types.md). Learn about the steps to migrate an existing codebase to make use of nullable reference types in the [migrating an application to use nullable reference types tutorial](../tutorials/upgrade-to-nullable-references.md).
```

```csharp
public static async System.Collections.Generic.IAsyncEnumerable<int> GenerateSequence()
{
    for (int i = 0; i < 20; i++)
    {
        await Task.Delay(100);
        yield return i;
    }
}
```

You would enumerate the sequence using the `await foreach` statement:

```csharp
await foreach (var number in GenerateSequence())
{
    Console.WriteLine(number);
}
```

You can try asynchronous streams yourself in our tutorial on [creating and consuming async streams](../tutorials/generate-consume-asynchronous-stream.md).

## Indices and ranges

Ranges and indices provide a succinct syntax for specifying subranges in an array, <xref:System.Span%601>, or <xref:System.ReadOnlySpan%601>.

> You can specify an index **from the end**. You specify **from the end** using the `^` operator. You are familiar with `array[2]` meaning the element "2 from the start". Now, `array[^2]` means the element "2 from the end". The index `^0` means "the end", or one past that last element.

Reviewer 3: Would be also useful to specify what index means "the last element". `^1`? 
And "...one past that last element" reads with difficulty: there is no one after the last element. However, I don't know how to rephrase it.

Author: I tried "the index following the last element".
